subprojects/core/src/main/java/org/gradle/initialization/DefaultGradlePropertiesController.java

```diff
@@ -112,135 +115,185 @@ public Map<String, String> getProperties() {
             return gradleProperties().getProperties();
         }
 
-        protected T gradleProperties() {
-            return state.gradleProperties();
-        }
-
-        protected void loadProperties(File dir, boolean setSystemProperties) {
-            state = state.loadProperties(dir, setSystemProperties);
+        private GradleProperties gradleProperties() {
+            return checkLoaded().gradleProperties;
         }
 
-        protected void unload() {
-            state = createNotLoadedState(id);
+        private LoadedBuildScopedState checkLoaded() {
+            LoadedBuildScopedState loaded = this.loaded;
+            if (loaded == null) {
+                throw new IllegalStateException(String.format("GradleProperties for %s have not been loaded yet.", buildId));
+            }
+            return loaded;
         }
 
-        protected abstract PropertiesState<T> createNotLoadedState(I id);
-    }
+        private void loadProperties(File buildRootDir, boolean setSystemProperties) {
+            LoadedBuildScopedState loaded = this.loaded;
+            if (loaded != null) {
+                if (loaded.buildRootDir.equals(buildRootDir)) {
```

Comment:
🤔 I understand this behavior was already there but I wonder if this wouldn't end up hiding some unwanted interactions